### DISCUSSION TITLE
scrambled order while calling readData from packet

### DISCUSSION THREAD

**Discussion by jinnkim0112** [ANLAB-KAIST/KENSv3#166] [Project Q&A 2. Transport Layer (TCP)] [answered]
I got a problem where ackNum part and seqNum part gets changed when a packet is arrived.

I wrote this code to illustrate. (all other code is just put as blank)
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/71919489/6225db7f-5583-416a-98a6-94796256aded)
upon running test-kens-handshake, TestAccept_Backlog1 will be run. Since the first packet has a sequence number of 0, when I get the packet correctly, tcpHead[4] through tcpHead[7] must be zero. However, I get this result.
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/71919489/0c77d0b2-216d-4869-b507-a1c072d92627)
First 4 numbers(that are separated with a period) are the port numbers of source and destination. Next four numbers should be the seq number, however they are not zero. Instead, the next four numbers which are ack numbers become zero. 
I have never changed code other than TCPAssignment.cpp. I even downloaded kens again to check. The only changed code is as illustrated on the top.
Why does this happen? How can I fix this?

**Comment by Pusnow** [comment]
> Since the first packet has a sequence number of 0, when I get the packet correctly, tcpHead[4] through tcpHead[7] must be zero

In the KENS's TCP implementation (communicates with your TCP implementation for grading) uses random initial sequence number, so it is normal to get non-zero initial sequence number.

You should handle non-zero initial sequence number like below:

https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/Documentation/rfc/rfc793.txt#L2029-L2043

**Reply by jinnkim0112** [reply]
Sorry, let me illustrate further. 
In the upper code, after removing the abort(); and inspecting the TestAccept_Backlog1.pcap via wireshark gives this result.
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/71919489/fe4532da-033f-4112-91ae-e573ee33ddf1)
You can see that clients indeed sent the packet with the sequence number 0.
First, I thought that the sequence number was indeed a random number, but after looking at wireshark, it seems that this is not the case.

**Reply by Pusnow** [reply, marked as answer]
The sequence numbers are "relative sequence numbers". See: https://wiki.wireshark.org/TCP_Relative_Sequence_Numbers

To view real absolute sequence numbers in tshark, try below:

```
tshark -o tcp.relative_sequence_numbers:FALSE  -r TestAccept_Backlog1.pcap 
```

**Reply by jinnkim0112** [reply]
I see. Thank you for the answer.
![image](https://github.com/ANLAB-KAIST/KENSv3/assets/71919489/28b75d70-d1ef-4101-838e-f08f8292ed96)
